### PR TITLE
core: swap bolt impl and enable configuring raft freelist sync behavior

### PR DESCRIPTION
**Pull request by shoenig**
This PR swaps the underlying BoltDB implementation from boltdb/bolt
to go.etc.io/bbolt. Applies to both server raft store and client data store.

In addition, the Server has a new configuration option for disabling
NoFreelistSync on the underlying database.

Freelist option: https://github.com/etcd-io/bbolt/blob/master/db.go#L81
Consul equivelent PR: https://github.com/hashicorp/consul/pull/11720

Closes https://github.com/hashicorp/nomad/issues/11775

### CODE REVIEW

File: website/content/docs/upgrade/upgrade-specific.mdx [hunk 1]
```diff
@@ -51,6 +51,26 @@ The volume staging directory for new CSI plugin tasks will now be
 mounted to the task's `NOMAD_TASK_DIR` instead of the
 `csi_plugin.mount_config`.
 
+#### Server Raft Database
+
+The server raft database in `raft.db` will be automatically migrated to a new
+underlying implementation provided by `go.etcd.io/bbolt`. Downgrading to a previous
+version of the server after upgrading it to Nomad 1.3 is not supported.
```
Comment (schmichael) [inline website/content/docs/upgrade/upgrade-specific.mdx:58]: Maybe we should be a little more prescriptive to help folks out with our No Going Back upgrades:

>  Like any Nomad upgrade it is recommended that you take a snapshot of your database prior to upgrading to use in case you need to downgrade.

Heavily inspired by https://www.consul.io/docs/upgrading/upgrade-specific#raft-storage-changes but removed the `if you expect that you will need to downgrade.` because who ever _expects_ they'll need to downgrade? :grimacing:

File: website/content/docs/upgrade/upgrade-specific.mdx [hunk 2]
```diff
@@ -51,6 +51,26 @@ The volume staging directory for new CSI plugin tasks will now be
 mounted to the task's `NOMAD_TASK_DIR` instead of the
 `csi_plugin.mount_config`.
 
+#### Server Raft Database
```
Comment (schmichael) [inline website/content/docs/upgrade/upgrade-specific.mdx:54]: We should mention the client as well... this is a bit wordy but would work:

```suggestion
#### Server Raft and Client State Databases
```

Content could go _after_ the server and be something like:

> Likewise Nomad client's store some local state in a bolt database that will be upgraded to the new implementation. Downgrading clients in place is not possible. The client must be drained and have its state directory removed before restarting with an old version.

**Review by schmichael — APPROVED** [review]
RIP Original Bolt. Stablest abandonware I've ever used.

Left some ideas on the docs but nothing to block on.

File: website/content/docs/upgrade/upgrade-specific.mdx (hunk 2 above)
Comment (tgross) [reply website/content/docs/upgrade/upgrade-specific.mdx:54]: Just a heads up that we already have a note about client metadata upgrades in 1.3.0 because of https://github.com/hashicorp/nomad/pull/12078

Comment (shoenig) [reply website/content/docs/upgrade/upgrade-specific.mdx:54]: Yup, Tim's note already covers the client case

https://github.com/hashicorp/nomad/pull/12078/files#diff-7c21f2d55cea6dc7286b54d164df00955882370ed3dbdeccd055f89a48350cf2R28

File: website/content/docs/upgrade/upgrade-specific.mdx (hunk 1 above)
Comment (shoenig) [reply website/content/docs/upgrade/upgrade-specific.mdx]: :+1:

File: website/content/docs/upgrade/upgrade-specific.mdx (hunk 2 above)
Comment (schmichael) [reply website/content/docs/upgrade/upgrade-specific.mdx:54]: Fantastic! Sorry for the noise.
